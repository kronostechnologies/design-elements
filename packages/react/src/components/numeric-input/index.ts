export { NumericInput, type NumericInputProps } from './numeric-input';
