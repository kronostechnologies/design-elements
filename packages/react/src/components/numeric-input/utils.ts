/**
 * We allow to input incomplete number that is obvious. It's similar to the behavior of Number(value).
 * The difference is we can keep all digits (zero ending decimals).
 */
export function cleanIncompleteNumber(inputValue: string): string {
    let value = inputValue;

    if (value === '' || Number.isNaN(Number(value))) {
        return '';
    }

    if (value.endsWith('.')) {
        value = value.slice(0, -1);
    }

    // Add missing integral zero (.25 => 0.25)
    value = value.replace(/^(-?)\./, '$10.');

    return value;
}

export function isWithinPrecision(value: string, precision: number): boolean {
    if (precision === 0 && value.includes('.')) {
        return false;
    }
    if (precision > 0 && value.split('.')[1]?.length > precision) {
        return false;
    }
    return true;
}

export function cleanPastedContent(content: string): string {
    // Remove spaces and invisible characters (ex: \r, \n, ZWSP) around the copied text
    return content.trim();
}

export function truncateAtPrecision(value: string, precision: number): string {
    if (value === '' || !value.includes('.')) {
        return value;
    }

    const atDot = value.indexOf('.');
    return value.slice(0, precision === 0 ? atDot : atDot + precision + 1);
}

/**
 * Validate that value can be put in the input (including partially incomplete number)
 */
export function isValidValueForInput(value: string): boolean {
    if (value.includes('.')) {
        return /^-?\d*\.\d*$/.test(value);
    }
    return /^-?\d*$/.test(value);
}
