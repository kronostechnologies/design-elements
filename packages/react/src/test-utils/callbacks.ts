export function doNothing(): void {
    // do nothing
}
