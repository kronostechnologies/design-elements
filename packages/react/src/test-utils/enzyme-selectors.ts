import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
