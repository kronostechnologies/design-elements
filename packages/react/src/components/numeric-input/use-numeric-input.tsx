import {
    ChangeEvent,
    ChangeEventHandler,
    ClipboardEvent,
    ClipboardEventHandler,
    FocusEvent,
    FocusEventHandler,
    FormEventHandler,
    useCallback,
    useState,
} from 'react';
import { isNumber, isWithinPrecision, truncateAtPrecision } from '../../utils/math';
import { useTranslation } from '../../i18n/use-translation';
import {
    cleanIncompleteNumber,
    cleanPastedContent,
    isValidValueForInput,
} from './utils';

interface NumericInputCallbackData {
    value: string;
    valueAsNumber: number | null;
}

export interface UseNumericInputParams {
    defaultValue?: number | string;
    max?: number;
    min?: number;
    onBlur?(event: FocusEvent<HTMLInputElement>, data: NumericInputCallbackData): void;
    onChange?(event: ChangeEvent<HTMLInputElement>, data: NumericInputCallbackData): void;
    precision?: number;
    required?: boolean;
    invalid?: boolean;
    validationErrorMessage?: string;
    value?: number | string;
}

export interface UseNumericInputReturn {
    max?: number;
    min?: number;
    onBlurHandler: FocusEventHandler<HTMLInputElement>;
    onChangeHandler: ChangeEventHandler<HTMLInputElement>;
    onPasteHandler: ClipboardEventHandler<HTMLInputElement>;
    onInvalid?: FormEventHandler<HTMLInputElement>;
    precision?: number;
    required?: boolean;
    invalid: boolean;
    validationErrorMessage?: string;
    value: number | string;
}

function validateProvidedValue(newValue: number | string): string {
    if (isNumber(newValue)) {
        return newValue.toString();
    }
    if (typeof newValue === 'string' && isValidValueForInput(newValue)) {
        return newValue;
    }

    console.warn(`Invalid value passed to NumericInput: ${newValue}`);
    return '';
}

type ValidationErrorType = 'required' | 'max' | 'min';

interface GetValidationErrorTypeOptions {
    max?: number;
    min?: number;
    required?: boolean;
    skipRequired?: boolean;
}

function getValidationErrorType(value: string, {
    max, min, required, skipRequired,
}: GetValidationErrorTypeOptions): ValidationErrorType | undefined {
    if (value === '' && required && !skipRequired) {
        return 'required';
    }
    if (value !== '') {
        if (max !== undefined && Number(value) > max) {
            return 'max';
        }
        if (min !== undefined && Number(value) < min) {
            return 'min';
        }
    }
    return undefined;
}

const createCallbackData = (inputValue: string): NumericInputCallbackData => ({
    value: inputValue,
    valueAsNumber: inputValue === '' ? null : Number(inputValue),
});

export function useNumericInput({
    defaultValue,
    max,
    min,
    onBlur,
    onChange,
    precision,
    required,
    invalid: providedInvalid,
    validationErrorMessage: providedValidationErrorMessage,
    value: providedValue,
}: UseNumericInputParams): UseNumericInputReturn {
    const { t } = useTranslation('numeric-input');
    const [inputValue, setInputValue] = useState(
        validateProvidedValue(providedValue ?? defaultValue ?? ''),
    );

    const [validationErrorType, setValidationErrorType] = useState<ValidationErrorType | undefined>(
        getValidationErrorType(inputValue, {
            max, min, required, skipRequired: true,
        }),
    );

    const validate = useCallback((value: string): void => {
        const errorType = getValidationErrorType(value, { max, min, required });
        setValidationErrorType(errorType);
    }, [max, min, required]);

    const handlePaste = useCallback((event: ClipboardEvent<HTMLInputElement>): void => {
        event.preventDefault();
        let newValue = event.clipboardData.getData('text/plain');

        newValue = cleanPastedContent(newValue);

        if (precision !== undefined) {
            newValue = truncateAtPrecision(newValue, precision);
        }

        // When the content is invalid, we will clear the field so the user has a feedback that the paste didn't work
        if (!isValidValueForInput(newValue)) {
            newValue = '';
        }

        setInputValue(newValue);
    }, [precision]);

    const handleChange = useCallback((event: ChangeEvent<HTMLInputElement>): void => {
        const value = event.target.value;
        if (value !== '') {
            if (!isValidValueForInput(value)) {
                return;
            }
            if (precision !== undefined && !isWithinPrecision(value, precision)) {
                return;
            }
        }

        setInputValue(value);
        onChange?.(event, createCallbackData(value));
    }, [onChange, precision]);

    const handleBlur = useCallback((event: FocusEvent<HTMLInputElement>): void => {
        const value = cleanIncompleteNumber(event.target.value);
        setInputValue(value);

        if (!(required && value === '')) {
            validate(value);
        } else {
            setValidationErrorType(undefined);
        }

        onBlur?.(event, createCallbackData(value));
    }, [onBlur, required, validate]);

    const handleOnInvalid: FormEventHandler<HTMLInputElement> = useCallback(() => {
        if (providedInvalid === undefined) {
            validate(inputValue);
        }
    }, [inputValue, providedInvalid, validate]);

    if (providedValue !== undefined) {
        const validatedProvidedValue = validateProvidedValue(providedValue);
        if (validatedProvidedValue !== inputValue) {
            setInputValue(validatedProvidedValue);
            validate(validatedProvidedValue);
        }
    }

    const getErrorMessage = (): string | undefined => {
        switch (validationErrorType) {
            case 'required':
                return t('requiredValidationErrorMessage');
            case 'max':
                return t('maxValidationErrorMessage', { max });
            case 'min':
                return t('minValidationErrorMessage', { min });
            default:
                return undefined;
        }
    };

    const isInvalid = providedInvalid ?? validationErrorType !== undefined;
    const errorMessage = providedValidationErrorMessage ?? getErrorMessage();

    return {
        max,
        min,
        onBlurHandler: handleBlur,
        onChangeHandler: handleChange,
        onPasteHandler: handlePaste,
        onInvalid: handleOnInvalid,
        precision,
        required,
        invalid: isInvalid,
        validationErrorMessage: errorMessage,
        value: inputValue,
    };
}
