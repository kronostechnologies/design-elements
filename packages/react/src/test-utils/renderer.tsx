import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
