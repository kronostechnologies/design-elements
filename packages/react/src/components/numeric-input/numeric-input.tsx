import {
    HTMLProps,
    ReactNode,
    useMemo,
    useRef,
    VoidFunctionComponent,
} from 'react';
import styled, { css } from 'styled-components';
import { useId } from '../../hooks/use-id';
import { focus } from '../../utils/css-state';
import { ResolvedTheme } from '../../themes/theme';
import { DeviceContextProps, useDeviceContext } from '../device-context-provider/device-context-provider';
import { FieldContainer } from '../field-container/field-container';
import { inputsStyle } from '../text-input/styles/inputs';
import { TooltipProps } from '../tooltip/tooltip';
import { useNumericInput, UseNumericInputParams } from './use-numeric-input';

interface StyledInputProps {
    device: DeviceContextProps;
    theme: ResolvedTheme;
    $textAlign: 'left' | 'right';
}

const StyledInput = styled.input<StyledInputProps>`
    ${({ theme, device }) => inputsStyle(theme, device.isMobile, false)}
    border: 0;
    flex: 1 1 auto;
    min-height: 100%;
    text-align: ${({ $textAlign }) => $textAlign};

    &:disabled {
        border: 0;
        box-shadow: none;
    }
`;

const Adornment = styled.span<{ $position: 'start' | 'end' }>`
    align-self: center;
    display: flex;
    padding-left: ${({ $position }) => ($position === 'start' ? 'var(--spacing-1x)' : undefined)};
    padding-right: ${({ $position }) => ($position === 'end' ? 'var(--spacing-1x)' : undefined)};
`;

interface StyledWrapperProps {
    $disabled?: boolean;
    $invalid?: boolean;
}

const Wrapper = styled.div<StyledWrapperProps>`
    background: ${({ theme }) => theme.component['numeric-input-background-color']};
    border: 1px solid ${({ theme }) => theme.component['numeric-input-border-color']};
    border-radius: var(--border-radius);
    box-sizing: border-box;
    display: flex;
    height: var(--size-2x);
    ${({ theme }) => focus({ theme }, false, undefined, true, false, false, 'focus-within')};
    ${({ $invalid, theme }) => $invalid && css`
        border-color: ${theme.component['numeric-input-invalid-border-color']};
`};
    ${({ $disabled, theme }) => $disabled && css`
        background-color: ${theme.component['numeric-input-disabled-background-color']};
        border-color: ${theme.component['numeric-input-disabled-border-color']};
        ${Adornment} {
            color: ${theme.component['numeric-input-disabled-adornment-text-color']};
        }
`};
`;

type NativeInputProps = Pick<HTMLProps<HTMLInputElement>, 'disabled' | 'onFocus' | 'placeholder'>;

interface NumericInputProps extends NativeInputProps {
    adornment?: ReactNode;
    adornmentPosition?: 'start' | 'end';
    className?: string;
    defaultValue?: number | string;
    hint?: string;
    id?: string;
    label?: string;
    max?: number;
    min?: number;
    noMargin?: boolean;
    onBlur?: UseNumericInputParams['onBlur'];
    onChange?: UseNumericInputParams['onChange'];
    precision?: number;
    required?: boolean;
    textAlign?: 'left' | 'right';
    tooltip?: TooltipProps;
    invalid?: boolean;
    validationErrorMessage?: string;
    value?: number | string;
}

export const NumericInput: VoidFunctionComponent<NumericInputProps> = ({
    adornment,
    adornmentPosition = 'start',
    className,
    defaultValue,
    disabled,
    hint,
    id,
    label,
    max,
    min,
    noMargin,
    onBlur,
    onChange,
    onFocus,
    placeholder,
    precision,
    required,
    textAlign = 'left',
    tooltip,
    invalid,
    validationErrorMessage,
    value,
}) => {
    const inputRef = useRef<HTMLInputElement>(null);
    const device = useDeviceContext();
    const fieldId = useId(id);

    const numericInput = useNumericInput({
        defaultValue,
        max,
        min,
        onBlur,
        onChange,
        precision,
        required,
        invalid,
        validationErrorMessage,
        value,
    });

    const adornmentContent = useMemo(() => (
        adornment ? <Adornment $position={adornmentPosition}>{adornment}</Adornment> : null
    ), [adornment, adornmentPosition]);

    return (
        <FieldContainer
            className={className}
            fieldId={fieldId}
            hint={hint}
            label={label}
            tooltip={tooltip}
            noMargin={noMargin}
            valid={!numericInput.invalid}
            noInvalidFieldIcon={!numericInput.validationErrorMessage}
            validationErrorMessage={numericInput.validationErrorMessage ?? ''}
        >
            <Wrapper $disabled={disabled} $invalid={numericInput.invalid}>
                {(adornment && adornmentPosition === 'start') && adornmentContent}
                <StyledInput
                    $textAlign={textAlign}
                    data-testid="numeric-input"
                    device={device}
                    disabled={disabled}
                    id={fieldId}
                    inputMode="numeric"
                    onBlur={numericInput.onBlurHandler}
                    onChange={numericInput.onChangeHandler}
                    onFocus={onFocus}
                    onPaste={numericInput.onPasteHandler}
                    placeholder={placeholder}
                    ref={inputRef}
                    type="text"
                    value={numericInput.value}
                />
                {(adornment && adornmentPosition === 'end') && adornmentContent}
            </Wrapper>
        </FieldContainer>
    );
};
