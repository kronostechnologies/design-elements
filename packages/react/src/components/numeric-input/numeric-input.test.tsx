import { shallow } from 'enzyme';
import { NumericInput } from './numeric-input';
import { getByTestId } from '../../test-utils/enzyme-selectors';
import { renderWithTheme } from '../../test-utils/renderer';

jest.mock('../../utils/uuid');

describe('NumericInput', () => {
    test('matches the snapshot (Normal - Adornment at start)', () => {
        const tree = renderWithTheme(<NumericInput adornment="%" value="50" />);

        expect(tree).toMatchSnapshot();
    });

    test('matches the snapshot (Normal - Adornment at end)', () => {
        const tree = renderWithTheme(<NumericInput adornment="%" adornmentPosition='end' value="50" />);

        expect(tree).toMatchSnapshot();
    });

    test('matches the snapshot (Disabled)', () => {
        const tree = renderWithTheme(<NumericInput adornment="%" value="50" disabled />);

        expect(tree).toMatchSnapshot();
    });

    test('matches the snapshot (Invalid)', () => {
        const tree = renderWithTheme(
            <NumericInput value="50" adornment="%" validationErrorMessage="This is an error message" invalid />,
        );

        expect(tree).toMatchSnapshot();
    });

    it('should call onChange with return value as number', () => {
        const onChange = jest.fn();
        const wrapper = shallow(
            <NumericInput onChange={onChange} />,
        );

        const event = { target: { value: '123.50' } };
        getByTestId(wrapper, 'numeric-input').invoke('onChange')(event);

        expect(onChange).toHaveBeenCalledWith(event, 123.5);
    });

    it('should call onBlur with return value as number', () => {
        const onBlur = jest.fn();
        const wrapper = shallow(
            <NumericInput onBlur={onBlur} />,
        );

        const event = { target: { value: '123.50' } };
        getByTestId(wrapper, 'numeric-input').invoke('onBlur')(event);

        expect(onBlur).toHaveBeenCalledWith(event, 123.5);
    });

    it('should call onChange with return value null when empty', () => {
        const onChange = jest.fn();
        const wrapper = shallow(
            <NumericInput onChange={onChange} />,
        );

        const event = { target: { value: '' } };
        getByTestId(wrapper, 'numeric-input').invoke('onChange')(event);

        expect(onChange).toHaveBeenCalledWith(event, null);
    });

    test('has controllable value', () => {
        const wrapper = shallow(<NumericInput value="500.25" />);

        expect(getByTestId(wrapper, 'numeric-input').prop('value')).toBe('500.25');
    });
});
