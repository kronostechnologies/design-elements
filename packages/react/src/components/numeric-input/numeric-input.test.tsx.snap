// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericInput matches the snapshot (Disabled) 1`] = `
.c0 {
  margin: 0 0 var(--spacing-3x);
}

.c0 input,
.c0 select,
.c0 textarea {
  border-color: #60666E;
}

.c0 > :nth-child(0) {
  margin-bottom: var(--spacing-quarter);
}

.c6 {
  background-color: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  color: #1B1C1E;
  font-family: inherit;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.015rem;
  -moz-letter-spacing: 0.015rem;
  -ms-letter-spacing: 0.015rem;
  letter-spacing: 0.015rem;
  line-height: 1.5rem;
  margin: 0;
  min-height: var(--size-2x);
  outline: none;
  padding: 0 var(--spacing-1x);
  width: 100%;
  border: 0;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  min-height: 100%;
}

.c6::-webkit-input-placeholder {
  color: #60666E;
}

.c6::-moz-placeholder {
  color: #60666E;
}

.c6:-ms-input-placeholder {
  color: #60666E;
}

.c6::placeholder {
  color: #60666E;
}

.c6:disabled {
  background-color: #F1F2F2;
  border-color: #B7BBC2;
  color: #B7BBC2;
}

.c6:disabled,
.c6:disabled::-webkit-input-placeholder {
  color: #B7BBC2;
}

.c6:disabled,
.c6:disabled::-moz-placeholder {
  color: #B7BBC2;
}

.c6:disabled,
.c6:disabled:-ms-input-placeholder {
  color: #B7BBC2;
}

.c6:disabled,
.c6:disabled::placeholder {
  color: #B7BBC2;
}

.c6:-moz-read-only {
  background-color: rgb(0 0 0 / 0.03);
  border-color: transparent;
  color: #1B1C1E;
}

.c6:read-only {
  background-color: rgb(0 0 0 / 0.03);
  border-color: transparent;
  color: #1B1C1E;
}

.c6:focus,
.c6:disabled {
  border: 0;
  box-shadow: none;
}

.c4 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  color: #60666E;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 > svg {
  height: 16px;
  width: 16px;
}

.c5 {
  padding-left: var(--spacing-1x);
}

.c2 {
  background-color: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: var(--size-2x);
  background-color: #F1F2F2;
  border-color: #B7BBC2;
}

.c2 {
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c2:focus-within {
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c2 .c3 {
  color: #B7BBC2;
}

.c1 .eds-TextInput-control {
  text-align: left;
}

.c1 .eds-TextInput-leftAdornment,
.c1 .eds-TextInput-rightAdornment {
  color: #1B1C1E;
}

<div
  class="c0 c1"
  data-testid="field-container"
>
  <div
    class="c2 eds-TextInput-control"
  >
    <span
      class="c3 c4 c5 eds-TextInput-leftAdornment"
    >
      %
    </span>
    <input
      class="c6"
      data-testid="numeric-input"
      disabled=""
      id="uuid1"
      inputmode="numeric"
      type="text"
      value="50"
    />
  </div>
</div>
`;

exports[`NumericInput matches the snapshot (Invalid) 1`] = `
.c2 {
  color: #CD2C23;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.75rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.25rem;
}

.c3 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: var(--spacing-base);
}

.c0 {
  margin: 0 0 var(--spacing-3x);
}

.c0 input,
.c0 select,
.c0 textarea {
  border-color: #CD2C23;
}

.c0 > :nth-child(1) {
  margin-bottom: var(--spacing-quarter);
}

.c7 {
  background-color: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  color: #1B1C1E;
  font-family: inherit;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.015rem;
  -moz-letter-spacing: 0.015rem;
  -ms-letter-spacing: 0.015rem;
  letter-spacing: 0.015rem;
  line-height: 1.5rem;
  margin: 0;
  min-height: var(--size-2x);
  outline: none;
  padding: 0 var(--spacing-1x);
  width: 100%;
  border: 0;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  min-height: 100%;
}

.c7::-webkit-input-placeholder {
  color: #60666E;
}

.c7::-moz-placeholder {
  color: #60666E;
}

.c7:-ms-input-placeholder {
  color: #60666E;
}

.c7::placeholder {
  color: #60666E;
}

.c7:disabled {
  background-color: #F1F2F2;
  border-color: #B7BBC2;
  color: #B7BBC2;
}

.c7:disabled,
.c7:disabled::-webkit-input-placeholder {
  color: #B7BBC2;
}

.c7:disabled,
.c7:disabled::-moz-placeholder {
  color: #B7BBC2;
}

.c7:disabled,
.c7:disabled:-ms-input-placeholder {
  color: #B7BBC2;
}

.c7:disabled,
.c7:disabled::placeholder {
  color: #B7BBC2;
}

.c7:-moz-read-only {
  background-color: rgb(0 0 0 / 0.03);
  border-color: transparent;
  color: #1B1C1E;
}

.c7:read-only {
  background-color: rgb(0 0 0 / 0.03);
  border-color: transparent;
  color: #1B1C1E;
}

.c7:focus,
.c7:disabled {
  border: 0;
  box-shadow: none;
}

.c5 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  color: #60666E;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c5 > svg {
  height: 16px;
  width: 16px;
}

.c6 {
  padding-left: var(--spacing-1x);
}

.c4 {
  background-color: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: var(--size-2x);
  border-color: #CD2C23;
}

.c4 {
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c4:focus-within {
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c1 .eds-TextInput-control {
  text-align: left;
}

.c1 .eds-TextInput-leftAdornment,
.c1 .eds-TextInput-rightAdornment {
  color: #1B1C1E;
}

<div
  class="c0 c1"
  data-testid="field-container"
>
  <span
    class="c2"
    data-testid="invalid-field"
    id="uuid1_invalid"
  >
    <svg
      class="c3"
      color="currentColor"
      focusable="false"
      height="16"
      width="16"
    />
    This is an error message
  </span>
  <div
    class="c4 eds-TextInput-control"
  >
    <span
      class="c5 c6 eds-TextInput-leftAdornment"
    >
      %
    </span>
    <input
      aria-describedby="uuid1_invalid"
      aria-invalid="true"
      class="c7"
      data-testid="numeric-input"
      id="uuid1"
      inputmode="numeric"
      type="text"
      value="50"
    />
  </div>
</div>
`;

exports[`NumericInput matches the snapshot (Normal - Adornment at end) 1`] = `
.c0 {
  margin: 0 0 var(--spacing-3x);
}

.c0 input,
.c0 select,
.c0 textarea {
  border-color: #60666E;
}

.c0 > :nth-child(0) {
  margin-bottom: var(--spacing-quarter);
}

.c3 {
  background-color: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  color: #1B1C1E;
  font-family: inherit;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.015rem;
  -moz-letter-spacing: 0.015rem;
  -ms-letter-spacing: 0.015rem;
  letter-spacing: 0.015rem;
  line-height: 1.5rem;
  margin: 0;
  min-height: var(--size-2x);
  outline: none;
  padding: 0 var(--spacing-1x);
  width: 100%;
  border: 0;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  min-height: 100%;
}

.c3::-webkit-input-placeholder {
  color: #60666E;
}

.c3::-moz-placeholder {
  color: #60666E;
}

.c3:-ms-input-placeholder {
  color: #60666E;
}

.c3::placeholder {
  color: #60666E;
}

.c3:disabled {
  background-color: #F1F2F2;
  border-color: #B7BBC2;
  color: #B7BBC2;
}

.c3:disabled,
.c3:disabled::-webkit-input-placeholder {
  color: #B7BBC2;
}

.c3:disabled,
.c3:disabled::-moz-placeholder {
  color: #B7BBC2;
}

.c3:disabled,
.c3:disabled:-ms-input-placeholder {
  color: #B7BBC2;
}

.c3:disabled,
.c3:disabled::placeholder {
  color: #B7BBC2;
}

.c3:-moz-read-only {
  background-color: rgb(0 0 0 / 0.03);
  border-color: transparent;
  color: #1B1C1E;
}

.c3:read-only {
  background-color: rgb(0 0 0 / 0.03);
  border-color: transparent;
  color: #1B1C1E;
}

.c3:focus,
.c3:disabled {
  border: 0;
  box-shadow: none;
}

.c4 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  color: #60666E;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 > svg {
  height: 16px;
  width: 16px;
}

.c5 {
  padding-right: var(--spacing-1x);
}

.c2 {
  background-color: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: var(--size-2x);
}

.c2 {
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c2:focus-within {
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c1 .eds-TextInput-control {
  text-align: left;
}

.c1 .eds-TextInput-leftAdornment,
.c1 .eds-TextInput-rightAdornment {
  color: #1B1C1E;
}

<div
  class="c0 c1"
  data-testid="field-container"
>
  <div
    class="c2 eds-TextInput-control"
  >
    <input
      class="c3"
      data-testid="numeric-input"
      id="uuid1"
      inputmode="numeric"
      type="text"
      value="50"
    />
    <span
      class="c4 c5 eds-TextInput-rightAdornment"
    >
      %
    </span>
  </div>
</div>
`;

exports[`NumericInput matches the snapshot (Normal - Adornment at start) 1`] = `
.c0 {
  margin: 0 0 var(--spacing-3x);
}

.c0 input,
.c0 select,
.c0 textarea {
  border-color: #60666E;
}

.c0 > :nth-child(0) {
  margin-bottom: var(--spacing-quarter);
}

.c5 {
  background-color: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  color: #1B1C1E;
  font-family: inherit;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.015rem;
  -moz-letter-spacing: 0.015rem;
  -ms-letter-spacing: 0.015rem;
  letter-spacing: 0.015rem;
  line-height: 1.5rem;
  margin: 0;
  min-height: var(--size-2x);
  outline: none;
  padding: 0 var(--spacing-1x);
  width: 100%;
  border: 0;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  min-height: 100%;
}

.c5::-webkit-input-placeholder {
  color: #60666E;
}

.c5::-moz-placeholder {
  color: #60666E;
}

.c5:-ms-input-placeholder {
  color: #60666E;
}

.c5::placeholder {
  color: #60666E;
}

.c5:disabled {
  background-color: #F1F2F2;
  border-color: #B7BBC2;
  color: #B7BBC2;
}

.c5:disabled,
.c5:disabled::-webkit-input-placeholder {
  color: #B7BBC2;
}

.c5:disabled,
.c5:disabled::-moz-placeholder {
  color: #B7BBC2;
}

.c5:disabled,
.c5:disabled:-ms-input-placeholder {
  color: #B7BBC2;
}

.c5:disabled,
.c5:disabled::placeholder {
  color: #B7BBC2;
}

.c5:-moz-read-only {
  background-color: rgb(0 0 0 / 0.03);
  border-color: transparent;
  color: #1B1C1E;
}

.c5:read-only {
  background-color: rgb(0 0 0 / 0.03);
  border-color: transparent;
  color: #1B1C1E;
}

.c5:focus,
.c5:disabled {
  border: 0;
  box-shadow: none;
}

.c3 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  color: #60666E;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c3 > svg {
  height: 16px;
  width: 16px;
}

.c4 {
  padding-left: var(--spacing-1x);
}

.c2 {
  background-color: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: var(--size-2x);
}

.c2 {
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c2:focus-within {
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c1 .eds-TextInput-control {
  text-align: left;
}

.c1 .eds-TextInput-leftAdornment,
.c1 .eds-TextInput-rightAdornment {
  color: #1B1C1E;
}

<div
  class="c0 c1"
  data-testid="field-container"
>
  <div
    class="c2 eds-TextInput-control"
  >
    <span
      class="c3 c4 eds-TextInput-leftAdornment"
    >
      %
    </span>
    <input
      class="c5"
      data-testid="numeric-input"
      id="uuid1"
      inputmode="numeric"
      type="text"
      value="50"
    />
  </div>
</div>
`;
